dll/win32/shell32/CUserNotification.cpp

```diff
@@ -0,0 +1,452 @@
+/*
+ * Copyright 2018 Hermes Belusca-Maito
+ *
+ * Pass on icon notification messages to the systray implementation
+ * in the currently running shell.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+#include "precomp.h"
+
+#include <mmsystem.h>
+#undef PlaySound
+
+WINE_DEFAULT_DEBUG_CHANNEL(shell_notify);
+
+
+/* Use Windows-compatible window callback message */
+#define WM_TRAYNOTIFY   (WM_USER + 100)
+
+/* Notification icon ID */
+#define ID_NOTIFY_ICON  0
+
+/* Balloon timers */
+#define ID_BALLOON_TIMEOUT      1
+#define ID_BALLOON_DELAYREMOVE  2
+#define ID_BALLOON_QUERYCONT    3
+#define ID_BALLOON_SHOWTIME     4
+
+#define BALLOON_DELAYREMOVE_TIMEOUT 250 // milliseconds
+
+
+CUserNotification::CUserNotification() :
+    m_hWorkerWnd(NULL),
+    m_hIcon(NULL),
+    m_uTimeout(15000),
+    m_uInterval(10000),
+    m_uContinuePoolInterval(0),
+    m_cRetryCount(0),
+    m_dwInfoFlags(0),
+    m_hRes(S_OK),
+    m_pqc(NULL)
+{
+}
+
+CUserNotification::~CUserNotification()
+{
+    /* If we have a notification window... */
+    if (m_hWorkerWnd)
+    {
+        /* ... remove the notification icon and destroy the window */
+        RemoveIcon();
+        ::DestroyWindow(m_hWorkerWnd);
+        m_hWorkerWnd = NULL;
+    }
+
+    /* Destroy our local icon copy */
+    if (m_hIcon)
+        ::DestroyIcon(m_hIcon);
+}
+
+VOID CUserNotification::RemoveIcon()
+{
+    NOTIFYICONDATAW nid = {0};
+
+    nid.cbSize = NOTIFYICONDATAW_V3_SIZE; // sizeof(nid);
+    nid.hWnd = m_hWorkerWnd;
+    nid.uID  = ID_NOTIFY_ICON;
```

Comment:
Nope, because you'are going to be sensitive to the actual layout of the structure, and then one would suggest to use the C99 syntax `{.cbSize = ... , .hWnd = ... }` etc. that is not recognized by all computers. Let not try to open a Pandora box.